src/Mvc/Mvc.ViewFeatures/src/ServerComponentSerializer.cs

```diff
@@ -0,0 +1,162 @@
+
+    // Serialization:
+    // For a given response, MVC renders one or more markers in sequence, including a descriptor for each rendered
+    // component containing the information described above.
+
+    // Deserialization:
+    // To prevent a client from rendering an infinite amount of components, we require clients to send all component
+    // markers in order. They can do so thanks to the sequence included in the marker.
+    // When we process a marker we do the following.
+    // * We unprotect the data-protected information.
+    // * We validate that the sequence number for the descriptor goes after the previous descriptor.
+    // * We compare the invocationId for the previous descriptor against the invocationId for the current descriptor to make sure they match.
+    // By doing this we achieve three things:
+    // * We ensure that the descriptor came from the server.
+    // * We ensure that a client can't just send an infinite amount of components to render.
+    // * We ensure that we do the minimal amount of work in the case of an invalid sequence of descriptors.
+    //
+    // For example:
+    // A client can't just send 100 component markers and force us to process them if the server didn't generate those 100 markers.
+    //  * If a marker is out of sequence we will fail early, so we process at most n-1 markers.
+    //  * If a marker has the right sequence but the invocation ID is different we will fail at that point. We know for sure that the
+    //    component wasn't render as part of the same response.
+    //  * If a marker can't be unprotected we will fail early. We know that the marker was tampered with and can't be trusted.
+    internal class ServerComponentSerializer
+    {
+        private static readonly object ComponentSequenceKey = new object();
+        private readonly ITimeLimitedDataProtector _dataProtector;
+
+        public ServerComponentSerializer(IDataProtectionProvider dataProtectionProvider) =>
+            _dataProtector = dataProtectionProvider
+                .CreateProtector(ServerComponentSerializationSettings.DataProtectionProviderPurpose)
+                .ToTimeLimitedDataProtector();
+
+        public ServerComponentMarker SerializeInvocation(HttpContext context, Type type, bool prerendered)
+        {
+            var (sequence, serverComponent) = CreateSerializedServerComponent(context, type);
+            return prerendered ? ServerComponentMarker.Prerendered(sequence, serverComponent) : ServerComponentMarker.NonPrerendered(sequence, serverComponent);
+        }
+
+        private ServerComponentInvocationSequence GetOrCreateInvocationIdentifier(HttpContext context)
+        {
+            if (!context.Items.TryGetValue(ComponentSequenceKey, out var sequence))
+            {
+                var newSequence = new ServerComponentInvocationSequence();
+                context.Items.Add(ComponentSequenceKey, newSequence);
```

Comment:
I'm worried about this still. How would this work when something like the error handling middleware is in use? Do we clear items?

Reply:
How do you propose we solve this?

I'm not sure what middleware you are referring to `UseStatusCodeWithReExecute`?. In general, re-executing the request is problematic no matter what approach you take.

You get the same services and the same httpcontext. Any 'scoped' service will not be clean and we can do two things. 
* Not support re-executing the request. (Relatively rare).
* Create a "Page" scope from MVC somewhere accessible to the HTML helper, like the ViewContext.
  * Whenever we render a top level page/view we "create" the scope and we track it as part of the items in the context or a scoped service.
  * If the request gets re-executed then it gets "reset".
 * If we use a scoped service for this then we can simply resolve it when we are going to render a page and initialize it.

Reply:
I solved this as described https://github.com/aspnet/AspNetCore/pull/13147#discussion_r314686256